src/parser/TripleComponent.h

```diff
@@ -179,8 +180,31 @@ class TripleComponent {
     }
   }
 
-  /// Human readable output. Is used for debugging, testing, and for the
-  /// creation of descriptors and cache keys.
+  // Same as the above, but also consider the given local vocabulary. If the
+  // string is neither in `vocabulary` nor in `localVocab`, it will be added to
+  // `localVocab`. Therefore, we get a valid `Id` in any case. The modifier is
+  // `&&` because in our uses of this method, the `TripleComponent` object is
+  // created solely to call this method and we want to avoid copying the
+  // `std::string` when passing it to the local vocabulary.
+  template <typename Vocabulary>
+  [[nodiscard]] Id toValueId(const Vocabulary& vocabulary,
+                             LocalVocab& localVocab) && {
+    std::optional<Id> id = toValueId(vocabulary);
+    if (!id) {
+      // If `toValueId` could not convert to `Id`, we have a string, which we
+      // look up in (and potentially add to) our local vocabulary.
+      AD_CONTRACT_CHECK(isString());
```

Comment:
I think "textbook example" is a bit pretentious here, since we just introduced these two variants and it seems we haven't achieved a common understanding yet of what each of them mean :-)

But I agree with you: given the documentation of `toValueId(const Vocabulary&)` function, there is no way to make this check fail by calling the method with a strange `Vocabulary`.

The example also shows that unless methods don't have very clear specificiations, it can be tricky or impossible to decide, which of the two checks is the "right" one.